language/move-binary-format/src/check_bounds.rs

```diff
@@ -367,9 +367,15 @@ impl<'a> BoundsChecker<'a> {
         check_bounds_impl(self.view.signatures(), code_unit.locals)?;
 
         let locals = self.get_locals(code_unit)?;
-        // Use saturating add for stability; range checked above
+        // Use saturating add for stability
         let locals_count = locals.len().saturating_add(parameters.len());
 
+        // Locals and parameters together must fit into LocalIndex (u8)
+        if locals_count > LocalIndex::MAX as usize {
```

Comment:
I think it should be `locals_count > LocalIndex::MAX as usize + 1`. For example, you could have 256 locals with the last one being referred to by index 255. 

There's also a dedicated error code for this: TOO_MANY_LOCALS.

For some reason the original check that was here disappeared...
```
       // check if the number of parameters + locals is less than u8::MAX
        let locals_count = locals.len() + parameters.len();

        if locals_count > (u8::MAX as usize) + 1 {
            return Err(append_err_info(
                VMStatus::new(StatusCode::TOO_MANY_LOCALS),
                IndexKind::FunctionDefinition,
                function_def_idx,
            ));
        }
```

Reply:
The check is actually still there, it is only in the calling function. Thanks for pointing out!

So the bug is actually this function in reference analysis:

```
   fn iter_locals(&self) -> impl Iterator<Item = LocalIndex> {
        0..self.num_locals as LocalIndex
    }
```

It is `0..0` if num_locals is 256... The iteration here should never happen in terms of `u8` but `usize`. I'm not sure whether we have other occasions of this bug, so I think the safest is to restrict the size as `255`. I removed the check I added and instead adapted this check.